### ISSUE TITLE
The recall rate is very low

### ISSUE THREAD

**Issue by Vayne0227** [ultralytics/ultralytics#14355]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello author, I trained my dataset in YOLOV8 and found that the Precision was high (0.83) but the recall was low (0.21). How should I solve this problem

**Comment by glenn-jocher** [comment]
@Vayne0227 hello,

Thank you for reaching out and providing details about your issue. It's great to hear that your model is achieving a high precision, but I understand that the low recall can be concerning. Here are a few steps you can take to address this:

1. **Check Confidence Threshold**: Sometimes, a high confidence threshold can lead to fewer detections, which might be causing the low recall. Try lowering the confidence threshold to see if it improves recall.
    ```python
    results = model.predict(source='path/to/your/data', conf=0.25)
    ```

2. **Data Augmentation**: Ensure that your training data is diverse and representative of the scenarios you expect the model to perform in. Data augmentation techniques like flipping, rotation, and scaling can help improve recall by making the model more robust.

3. **Class Imbalance**: If your dataset has an imbalance in the number of instances per class, this can affect recall. Consider techniques like oversampling the minority class or using class weights to balance the training process.

4. **Hyperparameter Tuning**: Experiment with different hyperparameters such as learning rate, batch size, and number of epochs. Sometimes, fine-tuning these parameters can lead to better recall.

5. **Validation and Test Sets**: Ensure that your validation and test sets are representative of the real-world scenarios where you plan to deploy the model. This can help in better understanding the recall performance.

6. **Model Architecture**: If the above steps do not yield significant improvements, consider experimenting with different model architectures or versions of YOLO that might be better suited for your specific dataset.

If you haven't already, please ensure you are using the latest version of the YOLOv8 package. This can sometimes resolve unexpected issues due to bug fixes and improvements in newer releases.

For a more detailed analysis, providing a minimum reproducible example would be very helpful. You can find guidelines on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to share any additional details or code snippets if you need further assistance. We're here to help! 😊

**Comment by Vayne0227** [comment]
> @Vayne0227 hello,
> 
> Thank you for reaching out and providing details about your issue. It's great to hear that your model is achieving a high precision, but I understand that the low recall can be concerning. Here are a few steps you can take to address this:
> 
> 1. **Check Confidence Threshold**: Sometimes, a high confidence threshold can lead to fewer detections, which might be causing the low recall. Try lowering the confidence threshold to see if it improves recall.
>    ```python
>    results = model.predict(source='path/to/your/data', conf=0.25)
>    ```
> 2. **Data Augmentation**: Ensure that your training data is diverse and representative of the scenarios you expect the model to perform in. Data augmentation techniques like flipping, rotation, and scaling can help improve recall by making the model more robust.
> 3. **Class Imbalance**: If your dataset has an imbalance in the number of instances per class, this can affect recall. Consider techniques like oversampling the minority class or using class weights to balance the training process.
> 4. **Hyperparameter Tuning**: Experiment with different hyperparameters such as learning rate, batch size, and number of epochs. Sometimes, fine-tuning these parameters can lead to better recall.
> 5. **Validation and Test Sets**: Ensure that your validation and test sets are representative of the real-world scenarios where you plan to deploy the model. This can help in better understanding the recall performance.
> 6. **Model Architecture**: If the above steps do not yield significant improvements, consider experimenting with different model architectures or versions of YOLO that might be better suited for your specific dataset.
> 
> If you haven't already, please ensure you are using the latest version of the YOLOv8 package. This can sometimes resolve unexpected issues due to bug fixes and improvements in newer releases.
> 
> For a more detailed analysis, providing a minimum reproducible example would be very helpful. You can find guidelines on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).
> 
> Feel free to share any additional details or code snippets if you need further assistance. We're here to help! 😊

Thank you for your analysis. I think the problem may lie in the imbalance of categories, as my dataset contains a large number of car categories and only a small number of people categories. How should I modify the code to balance the category weights. And I'm curious why category imbalance affects recall

**Comment by glenn-jocher** [comment]
Hello @Vayne0227,

Thank you for the additional context. Category imbalance can indeed have a significant impact on recall, especially for the minority classes. When the dataset is imbalanced, the model tends to be biased towards the majority class, which can lead to poor performance on the minority class, thus lowering recall.

To address this, you can modify the training process to account for class imbalance. Here are a few strategies you can implement:

1. **Class Weights**: Assign higher weights to the minority classes in the loss function. This can help the model pay more attention to the underrepresented classes.

    ```python
    # Example of setting class weights in the training configuration
    class_weights = [1.0, 5.0]  # Adjust these weights based on your dataset
    model.train(data='path/to/your/data.yaml', class_weights=class_weights)
    ```

2. **Oversampling**: Increase the number of samples for the minority class by duplicating them in the training dataset. This can help balance the dataset.

3. **Data Augmentation**: Apply data augmentation techniques more aggressively to the minority class to artificially increase its representation in the dataset.

4. **Focal Loss**: Consider using focal loss, which is designed to address class imbalance by down-weighting easy examples and focusing more on hard, misclassified examples.

    ```python
    # Example of using focal loss
    from ultralytics import YOLO
    from ultralytics.yolo.utils.loss import FocalLoss

    model = YOLO('yolov8n.yaml')
    model.loss = FocalLoss()
    model.train(data='path/to/your/data.yaml')
    ```

### Why Category Imbalance Affects Recall

Category imbalance affects recall because the model becomes biased towards the majority class. Recall measures the ability of the model to find all relevant instances of a class. When the model is trained on an imbalanced dataset, it may not learn to recognize the minority class well, leading to many false negatives (missed detections) for that class, thus lowering recall.

By implementing the strategies mentioned above, you can help the model learn to recognize the minority class better, improving recall.

Please ensure you are using the latest version of the YOLOv8 package to benefit from the latest features and bug fixes. If you encounter any issues, providing a minimum reproducible example can greatly help in diagnosing the problem. You can find guidelines on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to share any further details or questions you might have. We're here to assist! 😊

**Comment by tcoln** [comment]
yo

> Hello @Vayne0227,
> 
> Thank you for the additional context. Category imbalance can indeed have a significant impact on recall, especially for the minority classes. When the dataset is imbalanced, the model tends to be biased towards the majority class, which can lead to poor performance on the minority class, thus lowering recall.
> 
> To address this, you can modify the training process to account for class imbalance. Here are a few strategies you can implement:
> 
> 1. **Class Weights**: Assign higher weights to the minority classes in the loss function. This can help the model pay more attention to the underrepresented classes.
>    ```python
>    # Example of setting class weights in the training configuration
>    class_weights = [1.0, 5.0]  # Adjust these weights based on your dataset
>    model.train(data='path/to/your/data.yaml', class_weights=class_weights)
>    ```
> 2. **Oversampling**: Increase the number of samples for the minority class by duplicating them in the training dataset. This can help balance the dataset.
> 3. **Data Augmentation**: Apply data augmentation techniques more aggressively to the minority class to artificially increase its representation in the dataset.
> 4. **Focal Loss**: Consider using focal loss, which is designed to address class imbalance by down-weighting easy examples and focusing more on hard, misclassified examples.
>    ```python
>    # Example of using focal loss
>    from ultralytics import YOLO
>    from ultralytics.yolo.utils.loss import FocalLoss
>    
>    model = YOLO('yolov8n.yaml')
>    model.loss = FocalLoss()
>    model.train(data='path/to/your/data.yaml')
>    ```
> 
> ### Why Category Imbalance Affects Recall
> Category imbalance affects recall because the model becomes biased towards the majority class. Recall measures the ability of the model to find all relevant instances of a class. When the model is trained on an imbalanced dataset, it may not learn to recognize the minority class well, leading to many false negatives (missed detections) for that class, thus lowering recall.
> 
> By implementing the strategies mentioned above, you can help the model learn to recognize the minority class better, improving recall.
> 
> Please ensure you are using the latest version of the YOLOv8 package to benefit from the latest features and bug fixes. If you encounter any issues, providing a minimum reproducible example can greatly help in diagnosing the problem. You can find guidelines on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).
> 
> Feel free to share any further details or questions you might have. We're here to assist! 😊

Yolov8 doesn't support class_weights args in train function.
